tasks/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+                }
+                ClientFlowEvent::DataReceived { data } => {
+                    if let Some(data) = trickle_down(data, self.active_tasks_mut(), |task, data| {
+                        task.process_data(data)
+                    }) {
+                        return SchedulerEvent::Unsolicited(Response::Data(data));
+                    }
+                }
+                ClientFlowEvent::StatusReceived { status } => match StatusV2::from(status) {
+                    StatusV2::Untagged(body) => {
+                        if let Some(body) =
+                            trickle_down(body, self.active_tasks_mut(), |task, body| {
+                                task.process_untagged(body)
+                            })
+                        {
+                            return SchedulerEvent::Unsolicited(Response::Status(Status::from(
+                                StatusV2::Untagged(body),
+                            )));
+                        }
+                    }
+                    StatusV2::Bye(bye) => {
+                        if let Some(bye) =
+                            trickle_down(bye, self.active_tasks_mut(), |task, bye| {
+                                task.process_bye(bye)
+                            })
+                        {
+                            return SchedulerEvent::Unsolicited(Response::Status(Status::from(
+                                StatusV2::Bye(bye),
+                            )));
+                        }
+                    }
+                    StatusV2::Tagged(Tagged { tag, body }) => {
+                        let handle = {
+                            let (handle, _) = self
+                                .active_tasks
+                                .iter()
+                                .find(|(_, (tag_, _))| tag == *tag_)
+                                .unwrap();
+
+                            *handle
+                        };
+
+                        let (_, task) = self.active_tasks.remove(&handle).unwrap();
+
+                        let output = Some(task.process_tagged(body));
+
+                        return SchedulerEvent::TaskFinished(TaskToken { handle, output });
+                    }
+                },
+            }
+        }
+    }
+
+    fn active_tasks_mut(&mut self) -> impl Iterator<Item = &mut Box<dyn TaskAny>> {
+        self.active_tasks.values_mut().map(|(_, task)| task)
+    }
+}
+
+#[derive(Debug)]
+pub enum SchedulerEvent {
+    TaskFinished(TaskToken),
+    Shutdown,
+    Unsolicited(Response<'static>),
+}
+
+#[derive(Eq)]
+pub struct TaskHandle<T: Task> {
+    handle: ClientFlowCommandHandle,
+    _phantom: PhantomData<T>,
+}
+
+impl<T: Task> Debug for TaskHandle<T> {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        f.debug_struct("TaskHandle")
+            .field("handle", &self.handle)
+            .finish()
+    }
+}
```

Comment:
fyi: What we want here but Rust doesn't support is called perfect derive, e.g. see https://crates.io/crates/perfect-derive or the linked blog post https://smallcultfollowing.com/babysteps//blog/2022/04/12/implied-bounds-and-perfect-derive.

Reply:
Thanks for the link! I stumbled upon this in another project and learned that there are already crates that do what perfect derive would do. (Was it https://github.com/mcarton/rust-derivative?) But we needed a custom derive that wasn't supported as far as I remember. Should this come up a few times, we could evaluate if we want to use a helper crate.